test/e2e/network_segmentation.go

```diff
@@ -62,314 +62,319 @@ var _ = Describe("Network Segmentation", func() {
-						append(cmd, strings.Split(defRoute, " ")...)...)
-					Expect(err).NotTo(HaveOccurred())
-
-					// 2. Add a new default route to use default network interface
-					gatewayIP := iputils.NextIP(iputils.Network(podIP).IP)
-					_, err = e2ekubectl.RunKubectl(udnPod.Namespace,
-						append(ipCommand, "route", "add", "default", "via", gatewayIP.String(), "dev", "eth0")...)
-					Expect(err).NotTo(HaveOccurred())
-				}
-				// Now try to reach the host from the UDN pod
-				defaultPodHostIP := udnPod.Status.HostIPs
-				for _, hostIP := range defaultPodHostIP {
-					By("checking the UDN pod can't reach the host on IP " + hostIP.IP)
-					ping := "ping"
-					if utilnet.IsIPv6String(hostIP.IP) {
-						ping = "ping6"
-					}
-					Consistently(func() bool {
-						_, err := e2ekubectl.RunKubectl(udnPod.Namespace, "exec", udnPod.Name, "--",
-							ping, "-c", "1", "-W", "1", hostIP.IP,
-						)
-						return err == nil
-					}, 4*time.Second).Should(BeFalse())
-				}
-
-				By("asserting UDN pod can't reach default services via default network interface")
-				// route setup is already done, get kapi IPs
-				kapi, err := cs.CoreV1().Services("default").Get(context.Background(), "kubernetes", metav1.GetOptions{})
-				Expect(err).NotTo(HaveOccurred())
-				for _, kapiIP := range kapi.Spec.ClusterIPs {
-					By("checking the UDN pod can't reach kapi service on IP " + kapiIP)
-					Consistently(func() bool {
-						return connectToServerViaDefaultNetwork(udnPodConfig, kapiIP, int(kapi.Spec.Ports[0].Port)) != nil
-					}, 5*time.Second).Should(BeTrue())
-				}
 			},
-			Entry(
-				"with L2 dualstack primary UDN",
-				networkAttachmentConfigParams{
-					name:     nadName,
-					topology: "layer2",
-					cidr:     fmt.Sprintf("%s,%s", userDefinedNetworkIPv4Subnet, userDefinedNetworkIPv6Subnet),
-					role:     "primary",
-				},
-				*podConfig(
-					"udn-pod",
-					withCommand(func() []string {
-						return httpServerContainerCmd(port)
-					}),
-					withNodeSelector(map[string]string{nodeHostnameKey: workerOneNodeName}),
-				),
-			),
-			Entry(
-				"with L3 dualstack primary UDN",
-				networkAttachmentConfigParams{
-					name:     nadName,
-					topology: "layer3",
-					cidr:     fmt.Sprintf("%s,%s", userDefinedNetworkIPv4Subnet, userDefinedNetworkIPv6Subnet),
-					role:     "primary",
-				},
-				*podConfig(
-					"udn-pod",
-					withCommand(func() []string {
-						return httpServerContainerCmd(port)
-					}),
-					withNodeSelector(map[string]string{nodeHostnameKey: workerOneNodeName}),
-				),
-			),
+			Entry("NetworkAttachmentDefinitions", func(c networkAttachmentConfigParams) error {
+				netConfig := newNetworkAttachmentConfig(c)
+				nad := generateNAD(netConfig)
+				_, err := nadClient.NetworkAttachmentDefinitions(f.Namespace.Name).Create(context.Background(), nad, metav1.CreateOptions{})
+				return err
+			}),
+			Entry("UserDefinedNetwork", func(c networkAttachmentConfigParams) error {
```

Comment:
nice @ormergi I didnt know about this testing construct. Pretty cool.

Reply:
Its pretty new, introduced in Ginkgo v2.14 release :smiley: 

Although its very practical it could make tests hard to follow due to the additional wrapping overhead.
I guess it require restructure some tests in some way to make it easier to follow, I tried to avoid adding more this this PR though.